### PR TITLE
Allow random circuit generator to be seeded

### PR DESCRIPTION
**Pull request by kevinsung**

### CODE REVIEW

File: cirq/testing/random_circuit.py [hunk 1]
```diff
@@ -72,15 +76,23 @@ def random_circuit(qubits: Union[Sequence[ops.Qid], int],
...
+
     moments: List[ops.Moment] = []
     for _ in range(n_moments):
         operations = []
-        free_qubits = set(q for q in qubits)
+        free_qubits = set(qubits)
         while len(free_qubits) >= max_arity:
-            gate, arity = choice(tuple(gate_domain.items()))
-            op_qubits = sample(free_qubits, arity)
+            gate, arity = tuple(gate_domain.items())[prng.randint(
```
Comment (bryano) [inline cirq/testing/random_circuit.py]: Seems wasteful to keep recomputing the tuple and length, but probably doesn't matter.

**Comment by bryano** [comment]
LGTM after coverage addressed.

File: cirq/testing/random_circuit.py (hunk 1 above)
Comment (kevinsung) [reply cirq/testing/random_circuit.py]: Done.

File: cirq/testing/random_circuit_test.py [hunk 2]
```diff
@@ -67,3 +68,31 @@ def test_random_circuit(n_qubits: Union[int, Sequence[cirq.Qid]],
...
+                              random_state=seed)
+    circuit2 = random_circuit(qubits=20,
+                              n_moments=20,
+                              op_density=0.7,
+                              random_state=seed)
+    assert circuit1 == circuit2
+
+
+@pytest.mark.parametrize('seed', [randint(0, 2**32) for _ in range(10)])
+def test_random_circuit_reproducible_with_random_state(seed):
```
Comment (bryano) [inline cirq/testing/random_circuit_test.py]: Given that the parameters of these tests are the same, they should probably just be combined. Something like
```for s in (seed, np.random.RandomState(seed)):```

Comment (kevinsung) [reply cirq/testing/random_circuit_test.py]: I've combined the tests.

File: cirq/testing/random_circuit.py [hunk 3]
```diff
@@ -72,15 +76,24 @@ def random_circuit(qubits: Union[Sequence[ops.Qid], int],
     if n_qubits < 1:
         raise ValueError('At least one qubit must be specified.')
 
+    if isinstance(random_state, np.random.RandomState):
+        prng = random_state
+    else:
+        prng = np.random.RandomState(random_state)
```
Comment (mpharrigan) [inline cirq/testing/random_circuit.py]:
```python
if random_state is None:
  prng = np.random
```

It's important not to change the default behavior of using the global random number generator. As @Strilanc  pointed out in #2112 , you could potentially get into a situation where repeated calls to `np.random.RandomState(None)` will give you less-than-random random states. For example, if the default seed is the time, if you call it fast enough you'd seed with the same values

Comment (kevinsung) [reply cirq/testing/random_circuit.py]: Good point. I'll fix this.

**Pull request merged**